File: Deepgram/deepgram_handler/app/services/session_timers.py
"""Session timer management for voice calls.

Handles two failure modes:
1. Response timeout — LLM stalls after user speaks (dead air)
2. Idle caller — caller goes silent after agent finishes speaking

Uses asyncio timers and communicates through callbacks.
"""

from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from typing import Any, Awaitable, Callable

logger = logging.getLogger(__name__)

# Delay between injecting exit/goodbye message and actually hanging up,
# giving TTS time to finish speaking.
POST_EXIT_DELAY_S = 3.0


@dataclass
class SessionTimerCallbacks:
    inject_message: Callable[[str], Awaitable[None]]
    end_call: Callable[[], Awaitable[None]]
    log: Callable[[str], Any]


class SessionTimers:
    """Manages response timeout and idle caller detection timers.

    All timing is done via asyncio tasks so they integrate naturally
    with the event loop running the Deepgram bridge.
    """

    def __init__(self, config: dict, callbacks: SessionTimerCallbacks) -> None:
        self._config = config
        self._cb = callbacks

        self._response_reengage_handle: asyncio.TimerHandle | None = None
        self._response_exit_handle: asyncio.TimerHandle | None = None
        self._idle_prompt_handle: asyncio.TimerHandle | None = None
        self._idle_exit_handle: asyncio.TimerHandle | None = None

        self._idle_prompted: bool = False
        self._exiting: bool = False
        self._paused: bool = False

    @property
    def enabled(self) -> bool:
        return self._config.get("enabled", True)

    # ------------------------------------------------------------------
    # Response timeout chain
    # ------------------------------------------------------------------

    def on_user_spoke(self) -> None:
        """User spoke — start response timers, clear idle timers."""
        if not self.enabled or self._exiting or self._paused:
            return

        self._clear_response_timers()
        self._clear_idle_timers()
        self._idle_prompted = False

        reengage_ms = self._config.get("response_reengage_ms", 0)
        exit_ms = self._config.get("response_exit_ms", 0)

        loop = asyncio.get_running_loop()

        if reengage_ms > 0:
            self._response_reengage_handle = loop.call_later(
                reengage_ms / 1000,
                lambda: asyncio.ensure_future(self._fire_response_reengage()),
            )

        if exit_ms > 0:
            self._response_exit_handle = loop.call_later(
                exit_ms / 1000,
                lambda: asyncio.ensure_future(self._fire_response_exit()),
            )

        self._cb.log(
            f"[SessionTimers] on_user_spoke — response timers started "
            f"(reengage={reengage_ms}ms, exit={exit_ms}ms)"
        )

    def on_agent_started_speaking(self) -> None:
        """Agent started speaking — cancel response and idle timers.

        Idle timers must also be cleared because the agent speaking means the
        session is active.  Without this, a long LLM delay between a filler
        phrase's AgentAudioDone and the real response's AgentStartedSpeaking
        can cause the idle timer to fire mid-response.
        """
        if not self.enabled or self._exiting or self._paused:
            return
        had_response = self._response_reengage_handle is not None or self._response_exit_handle is not None
        had_idle = self._idle_prompt_handle is not None or self._idle_exit_handle is not None
        self._clear_response_timers()
        self._clear_idle_timers()
        self._idle_prompted = False
        if had_response or had_idle:
            self._cb.log("[SessionTimers] on_agent_started_speaking — response/idle timers cancelled")

    # ------------------------------------------------------------------
    # Idle caller detection chain
    # ------------------------------------------------------------------

    def on_user_started_speaking(self) -> None:
        """User started speaking — cancel idle timers (barge-in)."""
        if not self.enabled or self._exiting or self._paused:
            return
        had_timers = self._idle_prompt_handle is not None or self._idle_exit_handle is not None
        self._clear_idle_timers()
        self._idle_prompted = False
        if had_timers:
            self._cb.log("[SessionTimers] on_user_started_speaking — idle timers cancelled")

    def on_agent_audio_done(self) -> None:
        """Agent finished speaking — start idle timers."""
        if not self.enabled or self._exiting or self._paused:
            return
        if self._idle_prompted:
            self._cb.log("[SessionTimers] on_agent_audio_done — skipped (idle_prompted guard)")
            return

        self._clear_idle_timers()

        prompt_ms = self._config.get("idle_prompt_ms", 0)
        if prompt_ms > 0:
            loop = asyncio.get_running_loop()
            self._idle_prompt_handle = loop.call_later(
                prompt_ms / 1000,
                lambda: asyncio.ensure_future(self._fire_idle_prompt()),
            )
            self._cb.log(
                f"[SessionTimers] on_agent_audio_done — idle timer started (prompt={prompt_ms}ms)"
            )

    # ------------------------------------------------------------------
    # Timer fire handlers
    # ------------------------------------------------------------------

    async def _fire_response_reengage(self) -> None:
        if self._exiting:
            return
        self._cb.log("[SessionTimers] Response re-engage timeout — injecting message")
        msg = self._config.get("response_reengage_message", "")
        if msg:
            await self._cb.inject_message(msg)

    async def _fire_response_exit(self) -> None:
        if self._exiting:
            return
        self._exiting = True
        self._clear_response_timers()
        self._clear_idle_timers()
        self._cb.log("[SessionTimers] Response exit timeout — injecting exit message")
        msg = self._config.get("response_exit_message", "")
        try:
            if msg:
                await self._cb.inject_message(msg)
        except Exception:
            self._cb.log("[SessionTimers] Failed to inject exit message, proceeding with hangup")
        await asyncio.sleep(self._config.get("post_exit_delay_s", POST_EXIT_DELAY_S))
        await self._cb.end_call()

    async def _fire_idle_prompt(self) -> None:
        if self._exiting:
            return
        self._idle_prompted = True
        self._cb.log("[SessionTimers] Idle prompt — injecting message")
        msg = self._config.get("idle_prompt_message", "")
        if msg:
            await self._cb.inject_message(msg)

        exit_ms = self._config.get("idle_exit_ms", 0)
        if exit_ms > 0:
            loop = asyncio.get_running_loop()
            self._idle_exit_handle = loop.call_later(
                exit_ms / 1000,
                lambda: asyncio.ensure_future(self._fire_idle_exit()),
            )

    async def _fire_idle_exit(self) -> None:
        if self._exiting:
            return
        self._exiting = True
        self._clear_idle_timers()
        self._clear_response_timers()
        self._cb.log("[SessionTimers] Idle exit timeout — injecting exit message")
        msg = self._config.get("idle_exit_message", "")
        try:
            if msg:
                await self._cb.inject_message(msg)
        except Exception:
            self._cb.log("[SessionTimers] Failed to inject idle exit message, proceeding with hangup")
        await asyncio.sleep(self._config.get("post_exit_delay_s", POST_EXIT_DELAY_S))
        await self._cb.end_call()

    # ------------------------------------------------------------------
    # Pause / Resume (used during mute)
    # ------------------------------------------------------------------

    def pause(self) -> None:
        """Pause all timers (used during mute)."""
        self._paused = True
        self._clear_response_timers()
        self._clear_idle_timers()
        self._cb.log("[SessionTimers] paused")

    def resume(self) -> None:
        """Resume timer capability (used on unmute)."""
        self._paused = False
        self._cb.log("[SessionTimers] resumed")

    # ------------------------------------------------------------------
    # Cleanup
    # ------------------------------------------------------------------

    def _clear_response_timers(self) -> None:
        if self._response_reengage_handle:
            self._response_reengage_handle.cancel()
            self._response_reengage_handle = None
        if self._response_exit_handle:
            self._response_exit_handle.cancel()
            self._response_exit_handle = None

    def _clear_idle_timers(self) -> None:
        if self._idle_prompt_handle:
            self._idle_prompt_handle.cancel()
            self._idle_prompt_handle = None
        if self._idle_exit_handle:
            self._idle_exit_handle.cancel()
            self._idle_exit_handle = None

    def clear_all(self) -> None:
        """Cancel all timers and prevent further actions."""
        self._exiting = True
        self._clear_response_timers()
        self._clear_idle_timers()
        self._cb.log("[SessionTimers] clear_all — all timers cancelled")

File: extensions/voice-call/src/session-timers.ts
/**
 * Session Timers
 *
 * Per-session timer manager for detecting unresponsive agents and idle callers.
 * Wired into Deepgram event handlers in the media bridge.
 *
 * Two independent timer chains:
 *
 * **Response timeout** (LLM stalls):
 *   User speaks → start timers
 *   - responseReengageMs: inject re-engage message
 *   - responseExitMs: inject exit message → hangup
 *   agentStartedSpeaking at any point → cancel both (silent recovery)
 *   User speaks again → restart fresh cycle
 *
 * **Idle caller** (caller goes silent):
 *   agentAudioDone → start idle timer
 *   - idlePromptMs: inject prompt, set idlePrompted flag
 *   - idleExitMs after prompt: inject goodbye → hangup
 *   Any user speech → cancel, clear flag
 *   Guard: agentAudioDone from injected prompt does NOT restart idle timer
 */

import type { SessionTimerConfig } from "./config.js";

export interface SessionTimerCallbacks {
  injectAgentMessage: (msg: string) => void;
  endCall: () => Promise<void>;
  log: (msg: string) => void;
}

export class SessionTimers {
  private readonly config: SessionTimerConfig;
  private readonly callbacks: SessionTimerCallbacks;

  // Response timeout timers
  private responseReengageTimer: ReturnType<typeof setTimeout> | null = null;
  private responseExitTimer: ReturnType<typeof setTimeout> | null = null;

  // Idle caller timers
  private idlePromptTimer: ReturnType<typeof setTimeout> | null = null;
  private idleExitTimer: ReturnType<typeof setTimeout> | null = null;

  // Guard: prevent agentAudioDone from the idle prompt injection from restarting the idle timer
  private idlePrompted = false;

  // Track whether we're in the exit path to prevent further timer actions
  private exiting = false;

  constructor(config: SessionTimerConfig, callbacks: SessionTimerCallbacks) {
    this.config = config;
    this.callbacks = callbacks;
    callbacks.log(
      `[SessionTimers] Created — enabled=${config.enabled} responseReengageMs=${config.responseReengageMs} responseExitMs=${config.responseExitMs} idlePromptMs=${config.idlePromptMs} idleExitMs=${config.idleExitMs}`,
    );
  }

  /**
   * Called when the user speaks. Starts response timeout timers and clears idle timers.
   */
  onUserSpoke(): void {
    this.callbacks.log(
      `[SessionTimers] onUserSpoke — enabled=${this.config.enabled} exiting=${this.exiting}`,
    );
    if (!this.config.enabled || this.exiting) return;

    this.clearResponseTimers();
    this.clearIdleTimers();
    this.idlePrompted = false;

    // Start response re-engage timer
    if (this.config.responseReengageMs > 0) {
      this.responseReengageTimer = setTimeout(() => {
        if (this.exiting) return;
        this.callbacks.log("[SessionTimers] Response re-engage timeout — injecting message");
        this.callbacks.injectAgentMessage(this.config.responseReengageMessage);
      }, this.config.responseReengageMs);
    }

    // Start response exit timer
    if (this.config.responseExitMs > 0) {
      this.responseExitTimer = setTimeout(() => {
        if (this.exiting) return;
        this.exiting = true;
        this.clearResponseTimers();
        this.clearIdleTimers();
        this.callbacks.log("[SessionTimers] Response exit timeout — injecting exit message");
        this.callbacks.injectAgentMessage(this.config.responseExitMessage);
        setTimeout(() => {
          void this.callbacks.endCall();
        }, 3000);
      }, this.config.responseExitMs);
    }
  }

  /**
   * Called when the agent starts speaking. Clears response timers (silent recovery).
   */
  onAgentStartedSpeaking(): void {
    this.callbacks.log(
      `[SessionTimers] onAgentStartedSpeaking — enabled=${this.config.enabled} hasResponseTimers=${this.responseReengageTimer !== null || this.responseExitTimer !== null}`,
    );
    if (!this.config.enabled) return;
    this.clearResponseTimers();
  }

  /**
   * Called when the agent finishes sending audio. Starts idle caller timer.
   * Skipped if we just injected the idle prompt (idlePrompted guard).
   */
  onAgentAudioDone(): void {
    this.callbacks.log(
      `[SessionTimers] onAgentAudioDone — enabled=${this.config.enabled} exiting=${this.exiting} idlePrompted=${this.idlePrompted}`,
    );
    if (!this.config.enabled || this.exiting) return;

    // Guard: don't restart idle timer after the idle prompt itself
    if (this.idlePrompted) return;

    this.clearIdleTimers();

    if (this.config.idlePromptMs > 0) {
      this.idlePromptTimer = setTimeout(() => {
        if (this.exiting) return;
        this.idlePrompted = true;
        this.callbacks.log("[SessionTimers] Idle prompt timeout — injecting prompt");
        this.callbacks.injectAgentMessage(this.config.idlePromptMessage);

        // Start exit timer after prompt
        if (this.config.idleExitMs > 0) {
          this.idleExitTimer = setTimeout(() => {
            if (this.exiting) return;
            this.exiting = true;
            this.clearIdleTimers();
            this.clearResponseTimers();
            this.callbacks.log("[SessionTimers] Idle exit timeout — injecting exit message");
            this.callbacks.injectAgentMessage(this.config.idleExitMessage);
            setTimeout(() => {
              void this.callbacks.endCall();
            }, 3000);
          }, this.config.idleExitMs);
        }
      }, this.config.idlePromptMs);
    }
  }

  /**
   * Called when the user starts speaking. Clears idle timers and resets idlePrompted.
   */
  onUserStartedSpeaking(): void {
    this.callbacks.log(
      `[SessionTimers] onUserStartedSpeaking — enabled=${this.config.enabled} hasIdleTimers=${this.idlePromptTimer !== null || this.idleExitTimer !== null}`,
    );
    if (!this.config.enabled) return;
    this.clearIdleTimers();
    this.idlePrompted = false;
  }

  /**
   * Clear all timers. Called during cleanup (handleStop) or when exiting.
   */
  clearAll(): void {
    this.callbacks.log("[SessionTimers] clearAll called");
    this.exiting = true;
    this.clearResponseTimers();
    this.clearIdleTimers();
  }

  private clearResponseTimers(): void {
    if (this.responseReengageTimer) {
      clearTimeout(this.responseReengageTimer);
      this.responseReengageTimer = null;
    }
    if (this.responseExitTimer) {
      clearTimeout(this.responseExitTimer);
      this.responseExitTimer = null;
    }
  }

  private clearIdleTimers(): void {
    if (this.idlePromptTimer) {
      clearTimeout(this.idlePromptTimer);
      this.idlePromptTimer = null;
    }
    if (this.idleExitTimer) {
      clearTimeout(this.idleExitTimer);
      this.idleExitTimer = null;
    }
  }
}
